File: packages/java-sdk/examples/TurboSignConditionalFields.java
/**
 * Example: Conditional (IF/THEN) Fields
 *
 * A checkbox can control other fields so signers only see what applies to them:
 *   - Give a "checkbox" field a stable metadata via FieldMetadata.forFieldKey("...").
 *   - Give a dependent field FieldMetadata.forConditional(new FieldConditional(key, operator, action)).
 *       operator: "is_checked" | "is_not_checked"  -- when the rule fires.
 *       action:   "show"   (hidden until the rule fires)
 *                 "unlock" (visible but read-only until the rule fires).
 *
 * One checkbox can drive any number of dependent fields -- give them the same
 * controllingFieldKey. Uses createSignatureReviewLink (no emails are sent).
 */

package examples;

import com.turbodocx.TurboDocxClient;
import com.turbodocx.models.*;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Arrays;

public class TurboSignConditionalFields {
    public static void main(String[] args) {
        try {
            TurboDocxClient client = new TurboDocxClient.Builder()
                .apiKey(getEnv("TURBODOCX_API_KEY", "your-api-key-here"))
                .orgId(getEnv("TURBODOCX_ORG_ID", "your-org-id-here"))
                .senderEmail(getEnv("TURBODOCX_SENDER_EMAIL", "support@yourcompany.com"))
                .senderName(getEnv("TURBODOCX_SENDER_NAME", "Your Company Name"))
                .build();

            byte[] pdfFile = Files.readAllBytes(Paths.get("../../ExampleAssets/advanced-contract.pdf"));

            System.out.println("Creating a review link with conditional fields...");

            CreateSignatureReviewLinkResponse result = client.turboSign().createSignatureReviewLink(
                new CreateSignatureReviewLinkRequest.Builder()
                    .file(pdfFile)
                    .fileName("advanced-contract.pdf")
                    .documentName("Conditional Fields Demo")
                    .recipients(Arrays.asList(
                        new Recipient("John Doe", "john@example.com", 1)
                    ))
                    .fields(Arrays.asList(
                        // Controlling checkboxes -- each carries a stable fieldKey.
                        new Field.Builder().type("checkbox").recipientEmail("john@example.com").page(1).x(60).y(120).width(20).height(20)
                            .metadata(FieldMetadata.forFieldKey("request_changes")).build(),
                        new Field.Builder().type("checkbox").recipientEmail("john@example.com").page(1).x(60).y(300).width(20).height(20)
                            .metadata(FieldMetadata.forFieldKey("override_amount")).build(),
                        new Field.Builder().type("checkbox").recipientEmail("john@example.com").page(1).x(60).y(480).width(20).height(20)
                            .metadata(FieldMetadata.forFieldKey("consent")).build(),

                        // show + is_checked -- HIDDEN until "request_changes" is checked.
                        new Field.Builder().type("text").recipientEmail("john@example.com").page(1).x(120).y(120).width(260).height(40)
                            .metadata(FieldMetadata.forConditional(new FieldConditional("request_changes", "is_checked", "show"))).build(),
                        // ONE checkbox driving a SECOND dependent (same controllingFieldKey) -- a signature.
                        new Field.Builder().type("signature").recipientEmail("john@example.com").page(1).x(120).y(180).width(200).height(50)
                            .metadata(FieldMetadata.forConditional(new FieldConditional("request_changes", "is_checked", "show"))).build(),

                        // unlock + is_checked -- VISIBLE but locked until "override_amount" is checked.
                        new Field.Builder().type("text").recipientEmail("john@example.com").page(1).x(120).y(300).width(150).height(30).defaultValue("1000.00")
                            .metadata(FieldMetadata.forConditional(new FieldConditional("override_amount", "is_checked", "unlock"))).build(),

                        // show + is_not_checked -- a "please explain" box shown only while consent is WITHHELD.
                        new Field.Builder().type("text").recipientEmail("john@example.com").page(1).x(120).y(480).width(260).height(40)
                            .metadata(FieldMetadata.forConditional(new FieldConditional("consent", "is_not_checked", "show"))).build(),

                        // A normal required signature with no rule -- always visible, always required.
                        new Field.Builder().type("signature").recipientEmail("john@example.com").page(1).x(120).y(620).width(200).height(50).required(true).build()
                    ))
                    .build()
            );

            System.out.println("✅ Review link created!");
            System.out.println("Document ID: " + result.getDocumentId());
            System.out.println("Preview URL: " + result.getPreviewUrl());

            // Validation: a malformed rule (unknown operator/action, or a missing/empty
            //   controllingFieldKey) is rejected with HTTP 400 and code "InvalidConditionalRule".
            // Fail-open: a well-formed rule whose controllingFieldKey matches NO checkbox is NOT an
            //   error -- the dependent field stays visible/editable. Double-check your keys match.
        } catch (Exception error) {
            System.out.println("Error: " + error.getMessage());
        }
    }

    private static String getEnv(String key, String fallback) {
        String value = System.getenv(key);
        return (value != null && !value.isEmpty()) ? value : fallback;
    }
}

File: packages/go-sdk/examples/turbosign_conditional_fields.go
//go:build ignore
// +build ignore

// Example: Conditional (IF/THEN) Fields
//
// A checkbox can control other fields so signers only see what applies to them:
//   - Give a "checkbox" field a stable Metadata.FieldKey.
//   - Give a dependent field a Metadata.Conditional rule that references that key.
//       Operator: "is_checked" | "is_not_checked"  -- when the rule fires.
//       Action:   "show"   (hidden until the rule fires)
//                 "unlock" (visible but read-only until the rule fires).
//
// One checkbox can drive any number of dependent fields -- give them the same
// ControllingFieldKey. Uses CreateSignatureReviewLink (no emails are sent).

package main

import (
	"context"
	"fmt"
	"os"

	turbodocx "github.com/TurboDocx/SDK/packages/go-sdk"
)

func main() {
	client, err := turbodocx.NewClientWithConfig(turbodocx.ClientConfig{
		APIKey:      getEnv("TURBODOCX_API_KEY", "your-api-key-here"),
		OrgID:       getEnv("TURBODOCX_ORG_ID", "your-org-id-here"),
		SenderEmail: getEnv("TURBODOCX_SENDER_EMAIL", "support@yourcompany.com"),
		SenderName:  getEnv("TURBODOCX_SENDER_NAME", "Your Company Name"),
	})
	if err != nil {
		fmt.Printf("Error creating client: %v\n", err)
		return
	}

	pdfFile, err := os.ReadFile("../../ExampleAssets/advanced-contract.pdf")
	if err != nil {
		fmt.Printf("Error reading file: %v\n", err)
		return
	}

	fmt.Println("Creating a review link with conditional fields...")

	ctx := context.Background()
	result, err := client.TurboSign.CreateSignatureReviewLink(ctx, &turbodocx.CreateSignatureReviewLinkRequest{
		File:         pdfFile,
		FileName:     "advanced-contract.pdf",
		DocumentName: "Conditional Fields Demo",
		Recipients: []turbodocx.Recipient{
			{Name: "John Doe", Email: "john@example.com", SigningOrder: 1},
		},
		Fields: []turbodocx.Field{
			// Controlling checkboxes -- each carries a stable FieldKey.
			{Type: "checkbox", RecipientEmail: "john@example.com", Page: 1, X: 60, Y: 120, Width: 20, Height: 20, Metadata: &turbodocx.FieldMetadata{FieldKey: "request_changes"}},
			{Type: "checkbox", RecipientEmail: "john@example.com", Page: 1, X: 60, Y: 300, Width: 20, Height: 20, Metadata: &turbodocx.FieldMetadata{FieldKey: "override_amount"}},
			{Type: "checkbox", RecipientEmail: "john@example.com", Page: 1, X: 60, Y: 480, Width: 20, Height: 20, Metadata: &turbodocx.FieldMetadata{FieldKey: "consent"}},

			// show + is_checked -- HIDDEN until "request_changes" is checked.
			{Type: "text", RecipientEmail: "john@example.com", Page: 1, X: 120, Y: 120, Width: 260, Height: 40,
				Metadata: &turbodocx.FieldMetadata{Conditional: &turbodocx.FieldConditional{ControllingFieldKey: "request_changes", Operator: "is_checked", Action: "show"}}},
			// ONE checkbox driving a SECOND dependent (same ControllingFieldKey) -- a signature.
			{Type: "signature", RecipientEmail: "john@example.com", Page: 1, X: 120, Y: 180, Width: 200, Height: 50,
				Metadata: &turbodocx.FieldMetadata{Conditional: &turbodocx.FieldConditional{ControllingFieldKey: "request_changes", Operator: "is_checked", Action: "show"}}},

			// unlock + is_checked -- VISIBLE but locked until "override_amount" is checked.
			{Type: "text", RecipientEmail: "john@example.com", Page: 1, X: 120, Y: 300, Width: 150, Height: 30, DefaultValue: "1000.00",
				Metadata: &turbodocx.FieldMetadata{Conditional: &turbodocx.FieldConditional{ControllingFieldKey: "override_amount", Operator: "is_checked", Action: "unlock"}}},

			// show + is_not_checked -- a "please explain" box shown only while consent is WITHHELD.
			{Type: "text", RecipientEmail: "john@example.com", Page: 1, X: 120, Y: 480, Width: 260, Height: 40,
				Metadata: &turbodocx.FieldMetadata{Conditional: &turbodocx.FieldConditional{ControllingFieldKey: "consent", Operator: "is_not_checked", Action: "show"}}},

			// A normal required signature with no rule -- always visible, always required.
			{Type: "signature", RecipientEmail: "john@example.com", Page: 1, X: 120, Y: 620, Width: 200, Height: 50, Required: true},
		},
	})
	if err != nil {
		// A malformed rule is rejected here with a 400 and code "InvalidConditionalRule".
		fmt.Printf("Error: %v\n", err)
		return
	}

	fmt.Println("✅ Review link created!")
	fmt.Printf("Document ID: %s\n", result.DocumentID)
	fmt.Printf("Preview URL: %s\n", result.PreviewURL)

	// Fail-open: a well-formed rule whose ControllingFieldKey matches NO checkbox is NOT an
	// error -- the dependent field stays visible/editable. Double-check your keys match exactly.
}

func getEnv(key, fallback string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return fallback
}
